### ISSUE TITLE
why use a vpath? 

### ISSUE THREAD

**Issue by buzai** [IrosTheBeggar/mStream#82]
i think the file`s source should save in a database , and use a md5(filepath) function to get a hash as a key to search.
if source is not in local disk , it can save a url in db , also should has a type. eg (0=local disk. 1=url. 2=...)

**Comment by IrosTheBeggar** [comment]
Hey @buzai, good question.  I've been meaning to write to some documentation on vPaths, so here it goes.

Originally mStream worked by storing hashes and the absolute filepaths.  I had to change this when I added the multi-user functionality. When dealing with multiple users I had to assure two things.  First, mStream would prevent users from accessing eachothers files.  Secondly, mStream would prevent filepath conflicts.

In the early days, mStream would access files through the `media` endpoint.  So a user would use a URL like `media/artist1/single.mp3`.   When I added multiple users I realized that some users might have the same `artist1`directory.    I needed way for the server to distinguish these filepaths from eachother.  This is where vPaths come in.

Now userA has a url like `userA-files/artist1/single.mp3` while userB has a vPath like `userB-files/artist1/single.mp3`.  In this example, `userA-files` and `userB-files` are the vPaths and prevent path conflicts bethweeen users.

Ontop of that, vPaths offer an easy way to add security.  If userA tries to access `userB-files/artist1/single.mp3`, they will be denied.  The server checks that userA has access to the `userB-files` vPath.  This check will fail, and userA will be denied access to that URL.

vPaths are stored in the database, because I built the database to reflect URL structure.  Since 80% of the mStream code base revolves around streaming file through URLs, making the DB reflect the URL structure greatly reduces the amount of code needed to make mStream work. 

I hope that clarifies things. If you have any more questions let me know.
